cranelift/codegen/src/isa/x64/inst/emit.rs

```diff
@@ -1048,46 +1050,40 @@ pub(crate) fn emit(inst: &Inst, sink: &mut MachBuffer<Inst>) {
                 RexFlags::clear_w(),
             );
         }
-
-        Inst::XMM_RM_R {
+        Inst::XMM_MOV_RM_R {
+            op,
+            src: srcE,
+            dst: reg_g,
+        }
+        | Inst::XMM_RM_R {
             op,
             src: srcE,
             dst: reg_g,
         } => {
             let rex = RexFlags::clear_w();
-
-            let opcode = match op {
-                SseOpcode::Addss => 0x0F58,
-                SseOpcode::Subss => 0x0F5C,
-                SseOpcode::Mulss => 0x0F59,
-                SseOpcode::Divss => 0x0F5E,
-                SseOpcode::Sqrtss => 0x0F51,
-                _ => unimplemented!("XMM_RM_R opcode"),
+            let (prefix, opcode) = match op {
+                SseOpcode::Addss => (LegacyPrefix::_F3, 0x0F58),
+                SseOpcode::Andps => (LegacyPrefix::None, 0x0F54),
+                SseOpcode::Andnps => (LegacyPrefix::None, 0x0F55),
+                SseOpcode::Divss => (LegacyPrefix::_F3, 0x0F5E),
+                SseOpcode::Movaps => (LegacyPrefix::None, 0x0F28),
+                SseOpcode::Movd => (LegacyPrefix::_66, 0x0F6E),
```

Comment:
Why does this sse opcode need to be present for this `RM/R` mode, as well as in an above variant for `R/R` moves? I'm afraid this would mean that there are two possible Inst representations for `movd xmm0, rax`, for instance, which is something to avoid in general (it makes pattern-matching more complicated, for instance).

Reply:
Moves between GPRs and XMM regs (bit-preserving or with FP/Int conversions) don't fit cleanly in either the GPR-instruction or SSE*-instruction categories.  Maybe cleanest to create a new top level category, or maybe two (one that does conversions, one that just copies?) with their own opcode enumerations, and the usual RM->R structure.

Reply:
> Why does this sse opcode need to be present for this `RM/R` mode, as well as in an above variant for `R/R` moves? I'm afraid this would mean that there are two possible Inst representations for `movd xmm0, rax`, for instance, which is something to avoid in general (it makes pattern-matching more complicated, for instance).

Hi @bnjbvr ... It doesn't and #1838 removes the R_R. Perhaps I should just combine these two patches? The R_R was following the initial convention here but I thought these move instructions really should fit under a structure with "RM" in the name which would also by doing so would characterize these move instructions more consistently with how they are described in the arch guides (specifically the Op/En columns).

Reply:
> Moves between GPRs and XMM regs (bit-preserving or with FP/Int conversions) don't fit cleanly in either the GPR-instruction or SSE*-instruction categories. Maybe cleanest to create a new top level category, or maybe two (one that does conversions, one that just copies?) with their own opcode enumerations, and the usual RM->R structure.

Ok, thanks. Will think about this a little before the next push.

Reply:
@julian-seward1 I haven't addressed your comment in code yet and was thinking about it for a separate patch from this one if possible. In that patch we could create another enum called something like SseDtMixedOpcode (standing for sse data transfer mixed register opcode or something distinguishing like that) to show that the operand's registers aren't all the same category. However doing that seems to not do much more than organize the instructions at a finer granularity of SSE instruction which then opens the door for many more categories at that same granularity (for example: xmm only data transfer instructions, sse scalar arithmetic, packed arithmetic, sse comparison, sse conversion, etc) which maybe has the downfall of organizing a lot of new data structures that instead of simplifying, just creates more logic that is duplicated across lowering. I'd be happy to do that though as it does give more organization which in general feels like a good thing, but we also could just keep the one SSE instruction category which maybe simplifies things (distinguishing only from GPR, AVX, etc):

![image](https://user-images.githubusercontent.com/45402135/85326562-58fdd600-b482-11ea-93b3-e3bc9cd130c6.png)

Any more thoughts here?

Reply:
I agree with Johnnie here. In general, I even think we could move away from the Vcode Inst variant names being low-level descriptors of the kind of instructions, and could generally name them around what they do; the categories mentioned by @jlb6740 in the previous comment make a lot of sense to that effect: `xmm_data_transfer` instead of `XMM_Mov_RM_R`, `alu_arithmetic`, etc.

Reply:
My view is, the further you move the insn names away from the underlying structure of the insn set (eg `xmm_data_transfer`, etc) the harder it is to follow, since readers then have to guess or infer which set of insns `xmm_data_transfer` maps to.

> maybe has the downfall of organizing a lot of new data structures that instead of simplifying,

I'm not sure I understand what is meant here.  Do you mean, it creates a lot of new top level `Inst` cases?  To be concrete, all I was suggesting, for the cross-register-set insns, was adding two new `Inst` cases, somewhat like this:
```
new Inst case
  // XMM <-> GPR moves
  Inst::Mov_XMM_GPR {
      toXMM: bool,  // true => gpr->xmm, false => xmm->gpr
      gpr: Reg,     // src or dst
      xmm: Reg,     // dst or src
      is64: bool,   // 32 or 64 bit transfer?  We don't care about anything else
      laneNo: u8    // in the XMM; it may well be that this isn't even necessary
  }

new Inst case
  // Floating point conversions
  Inst::Cvt_XMM_GPR {
      gpr: Reg,        // src or dst
      xmm: Reg,        // dst or src
      how: SseFICvtOp, // what the conversion is
      rmode: RMode     // rounding mode to use, if relevant and
                       // if the insn supports direct specification of rounding modes
  }

  enum SSeFICvtOp {
     32UtoF32, // needs rounding
     32StoF32, // needs rounding
     32UtoF64, // no loss of precision
     32StoF64, // no loss of precision
     64StoF64, // needs rounding
     64UtoF64, // needs rounding
     // All the following involve rounding (RM)
     F32toU32,
     F32toS32,
     F64toU32,
     F64toU32,
     F64toS64,
     F64toU64,
  }

  enum RMode {
     Zero,     // round to zero,
     PlusInf,  // round to +Inf,
     MinusInf, // round to -Inf,
     Nearest   // round to nearest,
     MXCSR     // round per mxcsr.RM
  }
```
I'm not claiming that all of those insns exist or make sense.  But I do think these two would cover all xmm-gpr moves that we care about generating.  For example,
```
   CVTSD2SI
   is Cvt_XMM_GPR { gpr:_, xmm:_, how: F64toS64, rmode: MXCSR }

   CVTTSD2SI is the same, but with fixed rmode:
   is Cvt_XMM_GPR { gpr:_, xmm:_, how: F64toS64, rmode: Zero }
```
Plain XMM-to-XMM moves might want to stay in their own Inst:: case, since they have special significance to the register allocator.  I wasn't suggesting any further subdivision of `Inst`.

Reply:
> My view is, the further you move the insn names away from the underlying structure of the insn set (eg xmm_data_transfer, etc) the harder it is to follow, since readers then have to guess or infer which set of insns xmm_data_transfer maps to.

I kind of disagree with this. The useful information in `xmm_mov_rm_r` is really the part about `xmm` (this is for float registers) and about `mov` (this is for moves); the rest of it requires to be aware of the x86 jargon, and doesn't bring new information compared to what's is in the `Inst` variant struct fields. Also, in my opinion trying to enforce a 1:1 mapping on Inst name to the list of instructions that use it is the best way to make sure that comments and code stay **out of** sync over a long period of time. It also makes the code generating these Inst variants very hard to read, because this naming hides what's really interesting at the lowering level (= the semantics) by giving a preference to the implementation detail (= the encoding).

Reply:
@julian-seward1 I thought you were suggesting too to change the "enum SseOpcode" in x64/inst/args.rs but I guess that wasn't the case and you were only referring to the instruction enums?

> > maybe has the downfall of organizing a lot of new data structures that instead of simplifying,
> 
> I'm not sure I understand what is meant here. Do you mean, it creates a lot of new top level `Inst` cases?

Sorry, I wasn't clear here. I was wondering if having whichever extra structures needed ... including the ones you quickly put together above, adds more logic to consider or patch and to eventually get "out of sync" as @bnjbvr points out. The advantage that see is that it feels like it organizes things so that they are more orderly, but not sure if it truly does do that or if it just hurts readability by giving you that many more puzzle pieces to reason and figure out how the lowering is taking place. The more code to break, the more puzzle pieces (structs, match cases, etc) to consider, is what I meant by complexity. So while I agree with @bnjbvr who I think summed my thoughts better, I do wonder if there is more benefit to having this organization than I initially see and if it the long run we'd regret not having that finer granularity of lowering detail. Frankly, I'll be happy to implement either way ... and not sure who will make that call since there is a difference of opinion here, but I am thinking any change could be saved for a separate pr?

Reply:
I agree with @jlb6740, and I think that this is something we can easily revisit afterwards if it appears to be more convenient to split it up, so let's keep it as is in the meanwhile.